test/mpi/solver/solver.cpp

```diff
@@ -0,0 +1,551 @@
+    static void preprocess(
+        gko::matrix_data<value_type, global_index_type>& data)
+    {
+        gko::utils::make_hpd(data, 1.5);
+    }
+};
+
+
+struct Bicgstab : SimpleSolverTest<gko::solver::Bicgstab<solver_value_type>> {
+    static double tolerance() { return 300 * reduction_factor(); }
+};
+
+
+struct Ir : SimpleSolverTest<gko::solver::Ir<solver_value_type>> {
+    static void preprocess(
+        gko::matrix_data<value_type, global_index_type>& data)
+    {
+        gko::utils::make_hpd(data, 1.5);
+    }
+
+    static typename solver_type::parameters_type build(
+        std::shared_ptr<const gko::Executor> exec)
+    {
+        return SimpleSolverTest<gko::solver::Ir<solver_value_type>>::build(exec)
+            .with_solver(
+                gko::solver::Cg<value_type>::build()
+                    .with_criteria(
+                        gko::stop::Iteration::build()
+                            .with_max_iters(iteration_count())
+                            .on(exec),
+                        gko::stop::ResidualNorm<value_type>::build()
+                            .with_baseline(gko::stop::mode::absolute)
+                            .with_reduction_factor(2 * reduction_factor())
+                            .on(exec))
+                    .on(exec))
+            .with_relaxation_factor(0.9);
+    }
+};
+
+
+template <typename T>
+class Solver : public ::testing::Test {
+protected:
+    using Config = T;
+    using SolverType = typename T::solver_type;
+    using Mtx = typename T::dist_matrix_type;
+    using local_index_type = typename T::local_index_type;
+    using global_index_type = typename T::global_index_type;
+    using value_type = typename T::value_type;
+    using mixed_value_type = gko::next_precision<value_type>;
+    using Vec = typename T::dist_vector_type;
+    using LocalVec = typename T::non_dist_vector_type;
+    using MixedVec = typename T::mixed_dist_vector_type;
+    using MixedLocalVec = typename T::mixed_non_dist_vector_type;
+    using Part = typename T::partition_type;
+
+    Solver()
+        : ref(gko::ReferenceExecutor::create()),
+          comm(MPI_COMM_WORLD),
+          rand_engine(15)
+    {}
+
+    void SetUp()
+    {
+        ASSERT_EQ(comm.size(), 3);
+        init_executor(ref, exec, comm);
+        part = nullptr;
+    }
+
+    void TearDown()
+    {
+        if (exec != nullptr) {
+            ASSERT_NO_THROW(exec->synchronize());
+        }
+    }
+
+    std::shared_ptr<Mtx> gen_mtx(int num_rows, int num_cols, int min_cols,
+                                 int max_cols)
+    {
+        auto mapping =
+            gko::test::generate_random_array<gko::distributed::comm_index_type>(
+                num_rows,
+                std::uniform_int_distribution<
+                    gko::distributed::comm_index_type>(0, comm.size() - 1),
+                rand_engine, ref);
+        part = Part::build_from_mapping(ref, mapping, comm.size());
```

Comment:
Should there also be some scenarios with specific type of partition mappings? As a side question, what happens if a process has no index for example?

Reply:
Good idea. But I might not be able to add this before my vacation.

Reply:
It might actually lead to issues, because solvers with empty system matrices might not even enter the iteration loop, which would prevent processes to participate in the MPI reductions. This might lead to deadlocks.

Reply:
I've added tests with a partition that has an empty part and that actually did lead to a deadlock. I've fixed this, so thanks for bringing this up

Reply:
Great, thanks for checking and for fixing!